### PR TITLE
Feat/profiler lengths

### PR DESCRIPTION
**Pull request by dvilelaf**
## Proposed changes

- Track lists, dicts and sets length in the profiler, and outputs the longest ones. 

**Note**: I'm not sure we should merge this as if one of these collections grew a lot, we should be able to see almost the same information in the gc most common objects, but since I used this today to have a look at long objects, I leave it here for evaluation.

I have not included deques and async.io queues as we have very few of them and didn't want to overcomplicate the profiler. I checked for those manually.

## Fixes

n/a

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by dvilelaf** [comment]
Object_count_(gc-len)

```
Profiling details for current AEA process: 2022-05-02 14:32:11.994388
=============================================
Run time: 361.813951 seconds
Cpu time: 69.169924 seconds,
Cpu/Run time: 19.117539
Memory: 82.218164 MB [Peak 84.792994 MB]
Threads: 6  ['MainThread', 'Profiling', 'AsyncMultiplexer', 'ThreadPoolExecutor-0_0', 'DecisionMaker', 'ExecTimeout']
Objects present:
 * Message (present):  5
 * Dialogue (present):  0
 * DialogueLabel (present):  0
 * Handler (present):  5
 * Model (present):  12
 * Behaviour (present):  2
 * Skill (present):  1
 * Connection (present):  3
 * Contract (present):  
 * Protocol (present):  
Objects created:
 * Message (created):  6942
 * Dialogue (created):  6942
 * DialogueLabel (created):  7219
 * Handler (created):  5
 * Model (created):  12
 * Behaviour (created):  80
 * Skill (created):  1
 * Connection (created):  3
 * Contract (created):  37
 * Protocol (created):  5
Most common objects in garbage collector (excluding blacklisted):
 * Signature (gc-common):  305
 * Transaction (gc-common):  285
 * Timestamp (gc-common):  259
 * ConsensusVersion (gc-common):  257
 * PartSetHeader (gc-common):  257
 * BlockID (gc-common):  257
 * Header (gc-common):  25
 * Block (gc-common):  257
 * SimpleId (gc-common):  233
 * GeneratedProtocolMessageType (gc-common):  224
Longest objects in garbage collector:
 * dict[str, tzfile] (gc-len):  595
 * list[Block] (gc-len):  257
 * dict[builtin_function_or_method, tuple] (gc-len):  22
 * dict[MessageDescriptor, GeneratedProtocolMessageType] (gc-len):  220
 * dict[tuple, list] (gc-len):  100
 * list[RegistrationStartupRound] (gc-len):  78
 * list[BasePeriodState] (gc-len):  78
 * dict[str, DistInfoDistribution] (gc-len):  72
 * dict[int, function] (gc-len):  68
 * dict[str, RegisteredOption] (gc-len):  62
```

File: aea/helpers/profiling.py [hunk 1]
```diff
@@ -208,18 +219,44 @@ def get_profile_data(self) -> Dict:
         }
 
 
-def get_most_common_objects_in_gc(number: int = 15) -> List[Tuple[str, int]]:
+def get_object_type(obj: Any) -> str:
+    """Returns a string representation of an object's type"""
+    return str(
+        getattr(obj, "__class__", type(obj)).__name__
+    )  # not all objects have the __class__ attribute
```
Comment (Karrenbelt) [inline aea/helpers/profiling.py:226]: `__name__` is already a string. Remove comment or add it on a separate line above to keep the return a single line and hence more readable. More specifically: built-in types have no `__class__` attribute

File: aea/helpers/profiling.py [hunk 2]
```diff
@@ -208,18 +219,44 @@ def get_profile_data(self) -> Dict:
...
     lock.acquire()
     try:
         for obj in gc.get_objects():
-            object_type = str(
-                getattr(obj, "__class__", type(obj)).__name__
-            )  # not all objects have the __class__ attribute
+            object_type = get_object_type(obj)
             if object_type not in PROFILER_TYPE_BLACK_LIST:
                 object_count[object_type] += 1
+            if type(obj) in (dict, list, set) and len(obj) > 0:
```
Comment (Karrenbelt) [inline aea/helpers/profiling.py:240]: so any black listed dict, list and set you also want to analyse, correct? The preferred way of checking type is using the built-in `isinstance`: `if obj and isinstance(obj, (list, set, dict)):`

File: aea/helpers/profiling.py [hunk 3]
```diff
@@ -208,18 +219,44 @@ def get_profile_data(self) -> Dict:
...
+                if isinstance(obj, dict):
+                    first_key = list(obj.keys())[0]
+                    key_type = get_object_type(first_key)
+                    first_element = obj[first_key]
+                elif isinstance(obj, set):
+                    first_element = next(iter(obj))
+                elif isinstance(obj, list):
+                    first_element = obj[0]
+                else:
+                    continue
```
Comment (Karrenbelt) [inline aea/helpers/profiling.py:251]: in python a container (list, set or dict in this case) can contain more than a single type. I'm also not quite sure what you mean to achieve here, please shortly describe intention in doc string or comment

File: aea/helpers/profiling.py [hunk 4]
```diff
@@ -208,18 +219,44 @@ def get_profile_data(self) -> Dict:
...
+                content_type = get_object_type(first_element)
+                object_lengths[
+                    f"{object_type}[{str(key_type) + ', ' if key_type else ''}{content_type}]"
+                ] = obj.__len__()
```
Comment (Karrenbelt) [inline aea/helpers/profiling.py:255]:
```
key =  f"{object_type}[{str(key_type) + ', ' if key_type else ''}{content_type}"
object_lengths[key] = len(obj)
```

**Comment by Karrenbelt** [comment]
```
Note: I'm not sure we should merge this as if one of these collections grew a lot, we should be able to see almost the same information in the gc most common objects, but since I used this today to have a look at long objects, I leave it here for evaluation.
```

I don't see the use of tracking the length of containers. Can you provide a potential use case where this might be useful beyond the scope of tracking the size of objects as already done? One issue is that if we have a nested container, many levels deep, you will not be able to detect this using the current extension. I don't really see the benefit of tracking this

Furthermore, almost all custom objects have a `__dict__` and hence qualify as containers

**Pull request closed without merging**